tests/functional/test_install.py

```diff
@@ -1870,3 +1870,22 @@ def test_install_include_work_dir_pkg(script, data):
                         data.find_links, 'simple',
                         expect_stderr=True, cwd=pkg_path)
     assert 'Requirement already satisfied: simple' in result.stdout
+
+
+@pytest.mark.parametrize('package_name', ('simple-package', 'simple_package',
+                                          'simple.package'))
+def test_install_verify_package_name_normalization(script, package_name):
+    """
+    Test that install of a package again using a name which
+    normalizes to the original package name, is a no-op
+    since the package is already installed
+    """
+    pkg_path = create_test_package_with_setup(
+        script, name='simple-package', version='1.0')
+    result = script.pip('install', '-e', '.',
+                        expect_stderr=True, cwd=pkg_path)
+    assert 'Successfully installed simple-package' in result.stdout
+
+    result = script.pip('install', package_name)
+    assert 'Requirement already satisfied: {}'.format(
+        package_name) in result.stdout
```

Comment:
```suggestion
@pytest.mark.parametrize(
    'package_name',
    ['simple-package', 'simple_package', 'simple.package'],
)
@pytest.mark.parametrize(
    'install_name',
    ['simple-package', 'simple_package', 'simple.package'],
)
def test_install_verify_package_name_normalization(
    script,
    package_name,
    install_name,
):
    """
    Test that install of a package again using a name which
    normalizes to the original package name, is a no-op
    since the package is already installed
    """
    create_basic_wheel_for_package(script, package_name, '1.0')

    result = script.pip_install_local(install_name)
    assert 'Successfully installed {}'.format(install_name) in result.stdout

    result = script.pip_install_local(install_name)
    message = 'Requirement already satisfied: {}'.format(install_name)
    assert message in result.stdout
```

I think this should get all cases covered.

Reply:
So I tried out this test, and all 9 assertions failed with

```
ERROR: Could not find a version that satisfies the requirement <install_name> (from versions: none)
ERROR: No matching distribution found for simple_package
```

And I also see that `script.pip_install_local(...)` looks for packages in `tests/data/packages` , but we are creating the basic wheel somewhere else, perhaps in the pytest fixture environment, so is `pip_install_local` the correct way to install these packages?

Also is there an issue in my initial approach? Or can I modify it as follows

1. Creating a package with `package_name` with `create_test_package_with_setup`
2. Installing a package locally with `pip install .` 
3. Using `install_name` to install it again.

Reply:
It’s more desirable to use `create_basic_wheel_for_package()` since installing from wheel is much faster than a directory. Otherwise your approach is fine. I also didn’t really check whether mine work; feel free to stick to `script.pip('install')` 🙂

Reply:
I think it's not really clear to me why all 9 combinations are needed for `package_name, install_name` and why the suggested steps work to test those.

Or can I stick with my original approach with 3 combinations, 1 `package_name` and 3 `install_name` along with using `create_test_package_with_setup` and `script.pip('install')`

Reply:
After trying out the functions of all three package name variants (with dot, dash and underscore) I found that `canonicalize_name` converts all three packages names  to a name with dash, but `safe_name` keeps the name with a dot unchanged. It's also stated in the docstring of `safe_name`

```
Convert an arbitrary string to a standard distribution name
    Any runs of non-alphanumeric/. characters are replaced with a single '-'.
```

```
for n in ['simple-package', 'simple_package', 'simple.package']:
    print('{}, {}'.format(safe_name(n), canonicalize_name(n)))

# simple-package, simple-package
# simple-package, simple-package
# simple.package, simple-package
```
Given this info, all 9 combinations of tests might not work, unless we update the `safe_name` regex to be the same as `canonicalize_name`

Reply:
Also see https://github.com/pypa/pip/issues/8064 as to why we can't use a wheel to create a package name with a `-`

Reply:
So, if I understand this correctly, does this mean that if the package is named `fluent.logger`, `pip install fluent-logger` will *never* be able to show the correct result? Because `pkg_resources` would store the key as `fluent.logger`, but there’s no way to “uncanonicalise” `fluent-logger` to that key. This is sad :(

I guess one way we can fix this is to re-implement `get_distribution()` in pip. This is actually not that difficult; pip already has a function `get_installed_distributions()`, all we need is to build a dict out of that (and call `canonicalize_name` on keys), and look up packages with `canonicalize_name(req.name)`. This is probably a good thing in the long run as well, since we do want to eventually move away from `pkg_resources`, and we avoid calling `pkg_resources` functions would help refactoring.

Reply:
> So, if I understand this correctly, does this mean that if the package is named `fluent.logger`, `pip install fluent-logger` will _never_ be able to show the correct result? Because `pkg_resources` would store the key as `fluent.logger`, but there’s no way to “uncanonicalise” `fluent-logger` to that key. This is sad :(

If I understand correctly, what you are trying to say is that if we have a wheel ` simple.package-1.0-py2.py3-none-any.whl`, and first install that via `pip install simple.package-1.0-py2.py3-none-any.whl`, and the try `pip install simple-package`, ideally this installation should return with `Requirements already satisfied`, but instead it will return with `ERROR: Could not find a version that satisfies the requirement simple-package (from versions: none)` ?


> 
> I guess one way we can fix this is to re-implement `get_distribution()` in pip. This is actually not that difficult; pip already has a function `get_installed_distributions()`, all we need is to build a dict out of that (and call `canonicalize_name` on keys), and look up packages with `canonicalize_name(req.name)`. This is probably a good thing in the long run as well, since we do want to eventually move away from `pkg_resources`, and we avoid calling `pkg_resources` functions would help refactoring.

I think you are talking about replacing all  usages of `pkg_resources.get_distribution(name)` in the codebase with a custom implementation of `get_distribution()` as per your idea. Sure I will look into implementing that rather than the ` no_marker.name = canonicalize_name(no_marker.name)` which is mostly a patch without fixing the underlying issue which we have found out via this PR 😊

Reply:
> If I understand correctly, what you are trying to say is that if we have a wheel […]

Not a wheel, but a package in general. This shouldn’t be a problem for wheels (because it always normalise the name), only sdists and source directories (i.e. when we install from `setup.py`). pip is correct to not be able to find `simple.package-1.0-py2.py3-none-any.whl` (the name is invalid), but it already can find `simple.package-1.0.tar.gz` because it has logic to normalise the name when *looking for* a package. This package, however, will be recorded as `simple.package` when installed, and pip will not be able to correctly find it with `simple-package` (but it should).

Reply:
Yes you are right. I ran `pip install --no-cache-dir --no-index simple.package --find-links foo/` where `foo` had `simple.package-1.0.tar.gz`, and both `pip list` and `pip freeze` show the package as `simple.package==1.0`, which I guess is what you meant by `package, however, will be recorded as simple.package when installed`

Also when you say `pip will not be able to correctly find it with simple-package`, do you mean pip won't be able to find it via `pip show`? Because `pip show simple-package` and `pip show simple_package` both return the details of `simple.package` (perhaps because we canonicalize names of both the queries and the already installed packages in [pip._internal.commands.search_packages_info](https://github.com/pypa/pip/blob/master/src/pip/_internal/commands/show.py#L55) . Or did you mean something else?

Reply:
Also are we looking something like this for the new `get_distribution`

```
def get_distribution(req_name):
    req_name = canonicalize_name(req_name)
    packages = get_installed_distributions()
    pkg_dict = {canonicalize_name(p.key): p for p in packages}
    return pkg_dict.get(req_name)
```

A problem here is that `pkg_resources.get_distribution` also throws `DistributionNotFound` and `VersionConflict` exceptions, based on which the logic around the calls of `get_distribution` in `req_install` works, for e.g. https://github.com/pypa/pip/blob/master/src/pip/_internal/req/req_install.py#L430

 How would I catch those here and rethrow them to the caller to have the same effect as `pkg_resources.get_distribution` ?

Reply:
Hi @uranusjr 

It would be helpful if I could get clarifications on the questions above to move this PR ahead 😊

Reply:
> Because `pip show simple-package` and `pip show simple_package` both return the details of `simple.package`

The problem is more than `pip show`, but *all* things that need to find an already-installed distribution. I didn’t actually check, but IIRC all pip’s `pkg_resources.get_distribution` usages catch `DistributionNotFound` and/or `VersionConflict` anyway, so it’s easier to change those usages instead of replicating the exact behaviour of `pkg_resources`. In the `InstallRequirement` case you linked, for example, it’s as simple to do

```python
dist = get_distrubiton(self.req.name)  # Our own get_distribution.
if not dist:
    # Analogous to the except DistributionNotFound: block.
if self.req.specifier and dist.parsed_version in self.req.specifier:
    # Analogous to the else: block.
else:
    # Analogous to the except VersionConflict: block.
```

Reply:
Hi @uranusjr 

Thanks for the suggestions, let me try the code snippet you provided.

Another thing I found is that in some calls of `pkg_resources.get_distribution`, the `working_set` actually gets changed. (I saw this behaviour when running [tests.functional.test_uninstall:test_uninstall_setuptools_develop_install](https://github.com/pypa/pip/blob/master/tests/functional/test_uninstall.py#L506))

For example, the call at

https://github.com/pypa/pip/blob/a879bc448ae8555cb3e7c22457394e3647a0ee64/src/pip/_internal/req/req_install.py#L668

goes to [pkg_resources.get_provider](https://github.com/pypa/pkg_resources/blob/master/pkg_resources/__init__.py#L578) and then to  [pkg_resources.WorkingSet:resolve](https://github.com/pypa/pkg_resources/blob/6f81a44010d1266494025647dd1e1f0befa5b26b/pkg_resources/__init__.py#L1000-L1001), distributions gets added to the `working_set`.

How would be achieve this affect before our call to the custom `get_distribution` without using much of `pkg_resources` functionality?

Reply:
I’d say just ignore those, the current implementation looks good enough to me. `pkg_resources` contains a lot of old cruft that isn’t really relevant to modern packaging anymore, and IIUC even `importlib.metadata` does not implement those anyway.